### PR TITLE
glib: Simplify enum/flags class implementation

### PR DESCRIPTION
**Pull request by sdroege**
Less pointer wrapping and fewer allocations, and better `Debug` impl.

### CODE REVIEW

File: glib/src/enums.rs [hunk 1]
```diff
@@ -5,6 +5,8 @@ use crate::value::Value;
 use crate::CStr;
 use crate::Type;
 use std::cmp;
+use std::fmt;
+use std::ptr;
```
Comment (GuillaumeGomez) [inline glib/src/enums.rs:9]: Why not merging all these imports into one?

File: glib/src/enums.rs [hunk 2]
```diff
@@ -72,14 +83,15 @@ impl EnumClass {
...
+                ptr::NonNull::new(gobject_ffi::g_type_class_ref(type_.into_glib()) as *mut _)
+                    .unwrap(),
             ))
         }
     }
 
     /// `Type` of the enum.
     pub fn type_(&self) -> Type {
-        unsafe { from_glib((*self.0).g_type_class.g_type) }
+        unsafe { from_glib(self.0.as_ref().g_type_class.g_type) }
```
Comment (GuillaumeGomez) [inline glib/src/enums.rs:94]: That's super cool.

File: glib/src/enums.rs [hunk 3]
```diff
@@ -88,13 +100,13 @@ impl EnumClass {
...
-    pub fn value(&self, value: i32) -> Option<EnumValue> {
+    pub fn value(&self, value: i32) -> Option<&EnumValue> {
         unsafe {
-            let v = gobject_ffi::g_enum_get_value(self.0, value);
+            let v = gobject_ffi::g_enum_get_value(self.0.as_ptr(), value);
             if v.is_null() {
                 None
             } else {
-                Some(EnumValue(v, self.clone()))
+                Some(&*(v as *const EnumValue))
```
Comment (GuillaumeGomez) [inline glib/src/enums.rs:109]: No more clone! \o/

**Review by GuillaumeGomez — APPROVED** [review]
This is really great! Especially the ptr to slice conversions!

**Review by bilelmoussaoui — APPROVED** [review]

File: glib/src/enums.rs [hunk 4]
```diff
@@ -551,75 +575,79 @@ impl FlagsClass {
...
+                if v.value() & f != 0 {
+                    res.push(&*(v as *const FlagsValue));
                 }
             }
-            res
+            Some((flags_class, res))
         }
     }
-
-    /// Get `FlagsClass` to which the flags value belongs.
```
Comment (bilelmoussaoui) [inline glib/src/enums.rs:618]: I suppose it's fine to lose access to the FlagClass/EnumClass

Comment (sdroege) [reply glib/src/enums.rs:618]: Yes, you can't really have the value without the class.

**Pull request merged**